### PR TITLE
added effectivecouplings block

### PR DESCRIPTION
**Pull request by wkotlarski**

### CODE REVIEW

File: meta/WriteOut.m [hunk 1]
```diff
@@ -1227,6 +1227,10 @@
 {
    for (const auto& particle : decay_table) {
       set_decay_block(particle, flexibledecay_settings);
+      const int id = particle.get_particle_id();
+      if ((id % 25) % 10 == 0 || (id % 36) % 10 == 0 || (id % 37) % 10 == 0) {
```
Comment (Expander) [inline meta/WriteOut.m:1231]: I was just trying to wrap my head around this condition and could not fit it to the PDG numbers I know (in particular due to the `%10` operation. Would you mind adding a comment or giving the condition a name like (don't know if this is the correct name)
~~~.cpp
const bool is_higgs_boson = (id % 25) % 10 == 0 || (id % 36) % 10 == 0 || (id % 37) % 10 == 0;
~~~

Comment (wkotlarski) [reply meta/WriteOut.m:1231]: This was a quick test. I think the proper syntax is `(id - 25) % 10` etc :smiling_face_with_tear:

**Comment by wkotlarski** [comment]
The real problem is how to extract those amplitudes. They are computed by `Decay::calculate_amplitude` functions. Currently they are only called from inside of `Decay::get_partial_width`. `Decay::get_partial_width` are `const` co they cannot set a `Decay` class variable as a side effect. I see few solutions
1. `Decay::get_partial_width` now returns a width. We could return an `std::pair<double, optional<complex<double>>>`. I used optional because we want effective couplings from some `get_partial_width` not not from all.
2. Currently the signature of `get_partial_width` is
    ```cpp
    template <>    
    double CLASSNAME::get_partial_width<H, G, G>(    
        const context_base& context,    
        const typename field_indices<H>::type& in_idx,    
        const typename field_indices<G>::type& out1_idx,    
        const typename field_indices<G>::type& out2_idx) const
    ```
    we could change it to take also the amplitude such that it's computed somewhere else and then save as effective couplings and passed to also `get_partial_width`.

File: meta/WriteOut.m [hunk 2]
```diff
@@ -1225,8 +1225,20 @@
  */
 void " <> modelName <> "_slha_io::set_decays(const " <> modelName <> "_decay_table& decay_table, FlexibleDecay_settings const& flexibledecay_settings)
 {
+   auto is_higgs =
+      [](int id) {
+         bool is_whatever_higgs = false;
+         for (auto candidate : {25, 35, 37}) {
```
Comment (Expander) [inline meta/WriteOut.m:1231]: Sorry to be too pedantic, but I think this is an excellent candidate for `std::any_of`. :) We could write something along the lines
~~~.cpp
const int pids[] = { 25, 35, 37 }; // Higgs particle IDs
return std::any_of(std::begin(pids), std::end(pids), [&](){ /* predicate here */ });
~~~

**Comment by Expander** [comment]
> The real problem is how to extract those amplitudes. They are computed by `Decay::calculate_amplitude` functions. Currently they are only called from inside of `Decay::get_partial_width`. `Decay::get_partial_width` are `const` co they cannot set a `Decay` class variable as a side effect. I see few solutions
> 
>     1. `Decay::get_partial_width` now returns a width. We could return an `std::pair<double, optional<complex<double>>>`. I used optional because we want effective couplings from some `get_partial_width` not not from all.
> 
>     2. Currently the signature of `get_partial_width` is
>        ```c++
>        template <>    
>        double CLASSNAME::get_partial_width<H, G, G>(    
>            const context_base& context,    
>            const typename field_indices<H>::type& in_idx,    
>            const typename field_indices<G>::type& out1_idx,    
>            const typename field_indices<G>::type& out2_idx) const
>        ```
>        we could change it to take also the amplitude such that it's computed somewhere else and then save as effective couplings and passed to also `get_partial_width`.

Sorry, I don't understand the issue. Your patch seems fine to me. Should we discuss an alternative implementation?

**Comment by wkotlarski** [comment]
I've only setup an infrastructure to display effective couplings. For now I didn't yet figure out a way how to pass them from a point where they are calculated to the place that should display them.

**Comment by Expander** [comment]
Ok, I think I see the issue (please correct me if I'm wrong): In your current infrastructure you're passing the partial width to the function and you're currently writing the width out. However, you want to write amplitudes to the new effective couplings block, right?

Your option 1 sounds a bit fishy to me, because the function is called `get_partial_width` and the proposal is that it returns the partial width and something else, which is strange (and a bit confusing).

Your option 2 seems to pass things around, right? But somehow I don't fully understand how this is going to work, sorry. Just for my understanding: Which building blocks do you actually need to calculate the effective couplings?

**Comment by wkotlarski** [comment]
No problem. Currently there's a following chain
```c++
decays.set_decay(
   partial_width_hh_to_VPVP(dec_model.get()),
   {SM_info::get_pdg_code_for_particle(SM_info::VP), SM_info::get_pdg_code_for_particle(SM_info::VP)},
    create_process_string<hh, VP, VP>({},{},{})
);
```
where `partial_width_hh_to_VPVP` calls `get_partial_width<hh, VP, VP>` (`partial_width_hh_to_VPVP` is just a non-template wrapper over a templated function because Dylan wanted to have an extra layer) and finally `get_partial_width<hh, VP, VP>`  calls `calculate_amplitude<H, A, A>`. We could therefore do something like this
```cpp
const auto amp = calculate_amplitude<H, A, A>(context, in_idx, out1_idx, out2_idx);
decays.set_decay(
   partial_width_hh_to_VPVP(dec_model.get(), amp),
   {SM_info::get_pdg_code_for_particle(SM_info::VP), SM_info::get_pdg_code_for_particle(SM_info::VP)},
    create_process_string<hh, VP, VP>({},{},{})
);
decays.set_effectivecouplings(...);
```
The drawback is that this will make the code in `MODEL/decays/MODEL_decays.cpp` even longer. The part that was resolved by template arguments will now be written explicitly in the source code.

**Comment by wkotlarski** [comment]
Hi @Expander. I don't want to push you but I'd really like to finish this soon.

**Comment by Expander** [comment]
Sorry for the delay. Your suggestion seems to work fine. Although, I'm wondering if the extra `amp` argument to `partial_width_hh_to_VPVP` has any downsides. Shall it be an optional argument, so the function can be called w/o it?

**Comment by wkotlarski** [comment]
We could make it optional. And then only pass it to partial widths for which we also need effective couplings. For the rest we'd keep it as it is right now.

**Comment by Expander** [comment]
I think this would be good.

**Comment by wkotlarski** [comment]
The code structure is there. It works and produces reasonable output. I only need to test it on few models.

**Comment by wkotlarski** [comment]
@Expander, should the `EFFHIGGSCOUPLINGS` block be also given via the mathlink interface?

**Comment by Expander** [comment]
> @Expander, should the `EFFHIGGSCOUPLINGS` block be also given via the mathlink interface?

Do you mean whether it should appear in the output when using FS's `LibraryLink` interface, i.e. run it from within Mathematica? Yes, please! :)

**Comment by wkotlarski** [comment]
@Expander, @pathron - the code is ready for review. I have to check physics but this will not influence the code structure.

**Comment by Expander** [comment]
Looks good! Just asking: How does the LibraryLink output look like? Is it a list of replacement rules or similar?

**Comment by wkotlarski** [comment]
```mathematica
{
{CMSSM -> {hh[1] -> {25, 0.0031793597223488397, {{25, {-15, 15}, 0.00023862423125787935}, {25, {23, 23}, 0.000026246970915197293}, {25, {-24, 24}, 0.0002535577662306}, {25, {-3, 3}, 1.0275495839895232*^-6}, {25, {22, 22}, 6.850080467897071*^-6}, {25, {-13, 13}, 8.448505876066136*^-7}, {25, {22, 23}, 2.109519622325679*^-6}, {25, {-5, 5}, 0.0022672004091820706}, {25, {-4, 4}, 0.000106768632596377}, {25, {21, 21}, 0.00027612695467095516}}}, hh[2] -> {35, 1.4868385736423764, {{35, {-5, 5}, 0.9989819431293138}, {35, {-1000011, 1000011}, 0.0012940159147123258}, {35, {-1000012, 1000012}, 0.00021967519846103592}, {35, {21, 21}, 0.00028501376217154}, {35, {-15, 15}, 0.1465810627224004}, {35, {1000022, 1000023}, 0.016733961270479124}, {35, {-13, 13}, 0.0005182359357393949}, {35, {-3, 3}, 0.000455218340095155}, {35, {-1000016, 1000016}, 0.00019610171564832254}, {35, {-1000013, 1000013}, 0.00022228191642693565}, {35, {-1000013, 2000013}, 0.000015231557202118122}, {35, {-24, 24}, 0.0007917818439130839}, {35, {1000022, 1000022}, 0.006982085500839721}, {35, {-2000015, 1000011}, 0.0039785904932237515}, {35, {-1000015, 1000015}, 0.00021994055453779463}, {35, {23, 23}, 0.00040734834996336913}, {35, {-1000011, 2000015}, 0.0039785904932237515}, {35, {25, 25}, 0.005129931599490204}, {35, {-6, 6}, 0.29962371088020057}, {35, {-1000014, 1000014}, 0.00019619018939418193}, {35, {-2000013, 1000013}, 0.000015231557202118122}}}, Hpm[2] -> {-37, 1.2663349404703903, {{-37, {-24, 25}, 0.0008144452110528272}, {-37, {-4, 3}, 0.000389023048157312}, {-37, {-1000016, 2000011}, 0.00015596828797505134}, {-37, {-16, 15}, 0.1460490820288443}, {-37, {-1000012, 2000015}, 0.000035851516622578206}, {-37, {-14, 13}, 0.0005163420768520254}, {-37, {-1000014, 1000013}, 0.00003499272479975304}, {-37, {-1000012, 1000011}, 0.009758128812650577}, {-37, {-1000024, 1000022}, 0.07671773044940766}, {-37, {-6, 5}, 1.0317071972239986}, {-37, {-1000014, 2000013}, 0.00015536340524458534}}}, Ah[2] -> {36, 1.58555449252363, {{36, {-5, 5}, 0.9993783240325137}, {36, {-1000011, 2000015}, 0.004381312087314521}, {36, {1000022, 1000022}, 0.012307007011932648}, {36, {-15, 15}, 0.14661153453037606}, {36, {23, 25}, 0.0008053300221448652}, {36, {-24, 24}, 0.00003927617418517614}, {36, {-13, 13}, 0.000518330665214419}, {36, {-6, 6}, 0.354175833432113}, {36, {-3, 3}, 0.0004554607626564084}, {36, {1000022, 1000023}, 0.06199317154486755}, {36, {-2000015, 1000011}, 0.004381312087314521}, {36, {21, 21}, 0.0004470399345069444}}}}}, 
{CMSSM -> {{{25, 21, 21}, 0.00006764504668125032 + 3.0838369900683926*^-6*I}, {{25, 22, 22}, -0.000030165670984059015 + 2.2676693378749204*^-7*I}, {{25, 22, 23}, -0.00005265004437296775 + 2.0603189033073146*^-8*I}, {{35, 21, 21}, -2.3738827139174905*^-6 - 3.760149966956563*^-6*I}, {{35, 22, 22}, -8.815917542142452*^-7 - 1.00158199760603*^-6*I}, {{35, 22, 23}, -8.908980142456846*^-7 - 2.360399825338149*^-7*I}, {{36, 21, 21}, 0. + 4.157743354700155*^-26*I}, {{36, 22, 22}, -3.25557114345529*^-23 + 1.4171261380322486*^-23*I}, {{36, 22, 23}, 3.54475020486669*^-24 - 1.4740444358300086*^-22*I}}}
}
```

**Comment by Expander** [comment]
Very nice!
One request, though: Could you please add an example output to the documentation?

**Comment by wkotlarski** [comment]
Yes. I'll also add comment about using spectrum with MadGraph.

**Pull request closed without merging**